### PR TITLE
Contributed stubs for cachetools package

### PR DESCRIPTION
**Pull request by erictraut**

### CODE REVIEW

**Comment by hauntsaninja** [comment]
stubtest has a lot of complaints about this one, many of which seem to stem from extra arguments on base class methods

<details>

```
error: cachetools.Cache is not present in stub
Stub:
MISSING
Runtime: at line 15 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/cache.py
<class 'cachetools.cache.Cache'>

error: cachetools.LFUCache is not present in stub
Stub:
MISSING
Runtime: at line 6 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/lfu.py
<class 'cachetools.lfu.LFUCache'>

error: cachetools.LRUCache is not present in stub
Stub:
MISSING
Runtime: at line 6 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/lru.py
<class 'cachetools.lru.LRUCache'>

error: cachetools.RRCache is not present in stub
Stub:
MISSING
Runtime: at line 11 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/rr.py
<class 'cachetools.rr.RRCache'>

error: cachetools.TTLCache is not present in stub
Stub:
MISSING
Runtime: at line 55 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/ttl.py
<class 'cachetools.ttl.TTLCache'>

error: cachetools.cached is not present in stub
Stub:
MISSING
Runtime: at line 6 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/decorators.py
<function cached at 0x10b2513b0>

error: cachetools.cachedmethod is not present in stub
Stub:
MISSING
Runtime: at line 47 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/decorators.py
<function cachedmethod at 0x10bc7dc20>

error: cachetools.abc.DefaultMapping.get is inconsistent, runtime argument "default" has a default value of type None, which is incompatible with stub argument type Union[_VT_co`2, _T`-1]. This is often caused by overloads failing to account for explicitly passing in the default value.
Stub: at line 419
Overload(def (self: typing.Mapping[_KT`1, _VT_co`2], key: _KT`1) -> Union[_VT_co`2, None], def [_T] (self: typing.Mapping[_KT`1, _VT_co`2], key: _KT`1, default: Union[_VT_co`2, _T`-1]) -> Union[_VT_co`2, _T`-1])
Inferred signature: def (self: typing.Mapping[_KT`1, _VT_co`2], key: _KT`1, default: Union[_VT_co`2, _T`-1] = ...)
Runtime: at line 20 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/abc.py
def (self, key, default=None)

error: cachetools.cache.Cache.__delitem__ is inconsistent, runtime does not have argument "cache_delitem"
Stub: at line 11
def (cachetools.cache.Cache[_KT`1, _VT`2], _KT`1, def (_KT`1) =)
Runtime: at line 59 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/cache.py
def (self, key)

error: cachetools.cache.Cache.__getitem__ is inconsistent, runtime does not have argument "cache_getitem"
Stub: at line 9
def (cachetools.cache.Cache[_KT`1, _VT`2], _KT`1, def (_KT`1) -> _VT`2 =) -> _VT`2
Runtime: at line 37 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/cache.py
def (self, key)

error: cachetools.cache.Cache.__init__ is inconsistent, stub does not have argument "maxsize"
Stub: at line 98
def (self: builtins.object)
Runtime: at line 20 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/cache.py
def (self, maxsize, getsizeof=None)

error: cachetools.cache.Cache.__init__ is inconsistent, stub does not have argument "getsizeof"
Stub: at line 98
def (self: builtins.object)
Runtime: at line 20 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/cache.py
def (self, maxsize, getsizeof=None)

error: cachetools.cache.Cache.__setitem__ is inconsistent, runtime does not have argument "cache_setitem"
Stub: at line 10
def (cachetools.cache.Cache[_KT`1, _VT`2], _KT`1, _VT`2, def (_KT`1, _VT`2) =)
Runtime: at line 43 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/cache.py
def (self, key, value)

error: cachetools.cache.DefaultMapping.get is inconsistent, runtime argument "default" has a default value of type None, which is incompatible with stub argument type Union[_VT_co`2, _T`-1]. This is often caused by overloads failing to account for explicitly passing in the default value.
Stub: at line 419
Overload(def (self: typing.Mapping[_KT`1, _VT_co`2], key: _KT`1) -> Union[_VT_co`2, None], def [_T] (self: typing.Mapping[_KT`1, _VT_co`2], key: _KT`1, default: Union[_VT_co`2, _T`-1]) -> Union[_VT_co`2, _T`-1])
Inferred signature: def (self: typing.Mapping[_KT`1, _VT_co`2], key: _KT`1, default: Union[_VT_co`2, _T`-1] = ...)
Runtime: at line 20 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/abc.py
def (self, key, default=None)

error: cachetools.lru.Cache.__delitem__ is inconsistent, runtime does not have argument "cache_delitem"
Stub: at line 11
def (cachetools.cache.Cache[_KT`1, _VT`2], _KT`1, def (_KT`1) =)
Runtime: at line 59 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/cache.py
def (self, key)

error: cachetools.lru.Cache.__getitem__ is inconsistent, runtime does not have argument "cache_getitem"
Stub: at line 9
def (cachetools.cache.Cache[_KT`1, _VT`2], _KT`1, def (_KT`1) -> _VT`2 =) -> _VT`2
Runtime: at line 37 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/cache.py
def (self, key)

error: cachetools.lru.Cache.__init__ is inconsistent, stub does not have argument "maxsize"
Stub: at line 98
def (self: builtins.object)
Runtime: at line 20 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/cache.py
def (self, maxsize, getsizeof=None)

error: cachetools.lru.Cache.__init__ is inconsistent, stub does not have argument "getsizeof"
Stub: at line 98
def (self: builtins.object)
Runtime: at line 20 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/cache.py
def (self, maxsize, getsizeof=None)

error: cachetools.lru.Cache.__setitem__ is inconsistent, runtime does not have argument "cache_setitem"
Stub: at line 10
def (cachetools.cache.Cache[_KT`1, _VT`2], _KT`1, _VT`2, def (_KT`1, _VT`2) =)
Runtime: at line 43 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/cache.py
def (self, key, value)

error: cachetools.rr.Cache.__delitem__ is inconsistent, runtime does not have argument "cache_delitem"
Stub: at line 11
def (cachetools.cache.Cache[_KT`1, _VT`2], _KT`1, def (_KT`1) =)
Runtime: at line 59 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/cache.py
def (self, key)

error: cachetools.rr.Cache.__getitem__ is inconsistent, runtime does not have argument "cache_getitem"
Stub: at line 9
def (cachetools.cache.Cache[_KT`1, _VT`2], _KT`1, def (_KT`1) -> _VT`2 =) -> _VT`2
Runtime: at line 37 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/cache.py
def (self, key)

error: cachetools.rr.Cache.__init__ is inconsistent, stub does not have argument "maxsize"
Stub: at line 98
def (self: builtins.object)
Runtime: at line 20 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/cache.py
def (self, maxsize, getsizeof=None)

error: cachetools.rr.Cache.__init__ is inconsistent, stub does not have argument "getsizeof"
Stub: at line 98
def (self: builtins.object)
Runtime: at line 20 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/cache.py
def (self, maxsize, getsizeof=None)

error: cachetools.rr.Cache.__setitem__ is inconsistent, runtime does not have argument "cache_setitem"
Stub: at line 10
def (cachetools.cache.Cache[_KT`1, _VT`2], _KT`1, _VT`2, def (_KT`1, _VT`2) =)
Runtime: at line 43 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/cache.py
def (self, key, value)

error: cachetools.rr.RRCache.__delitem__ is inconsistent, runtime does not have argument "cache_delitem"
Stub: at line 17
def (cachetools.rr.RRCache[_KT`1, _VT`2], _KT`1, def (_KT`1) =)
Runtime: at line 59 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/cache.py
def (self, key)

error: cachetools.rr.RRCache.__getitem__ is inconsistent, runtime does not have argument "cache_getitem"
Stub: at line 15
def (cachetools.rr.RRCache[_KT`1, _VT`2], _KT`1, def (_KT`1) -> _VT`2 =) -> _VT`2
Runtime: at line 37 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/cache.py
def (self, key)

error: cachetools.rr.RRCache.__setitem__ is inconsistent, runtime does not have argument "cache_setitem"
Stub: at line 16
def (cachetools.rr.RRCache[_KT`1, _VT`2], _KT`1, _VT`2, def (_KT`1, _VT`2) =)
Runtime: at line 43 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/cache.py
def (self, key, value)

error: cachetools.ttl.TTLCache.expire is inconsistent, runtime argument "time" has a default value of type None, which is incompatible with stub argument type def () -> builtins.float
Stub: at line 23
def (self: cachetools.ttl.TTLCache[_KT`1, _VT`2], time: def () -> builtins.float =)
Runtime: at line 159 in file /Users/shantanu/.virtualenvs/typeshed/lib/python3.7/site-packages/cachetools/ttl.py
def (self, time=None)
```
</details>

**Comment by erictraut** [comment]
I fixed the missing symbols caught by stubtest, but I'm not sure how to interpret the other stubtest errors. They look incorrect to me, but maybe I'm missing something. If you see a fix, feel free to push directly to the branch.

**Comment by hauntsaninja** [comment]
Made the stubtest fixes! (The only stubtest complaint left is with `cachetools.abc.DefaultMapping.get` which is known to be invalid and I should get around to fixing)

**Comment by ikonst** [comment]
What is the recommended way of using these stubs?

The classes made generic here are not subscriptable in runtime:
```
>>> import cachetools
>>> cachetools.TTLCache
>>> cachetools.TTLCache[int, int]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'ABCMeta' object is not subscriptable
```

**Comment by hauntsaninja** [comment]
https://mypy.readthedocs.io/en/stable/runtime_troubles.html

File: third_party/2and3/cachetools/decorators.pyi [hunk 1]
```diff
@@ -0,0 +1,11 @@
+from typing import Any, Callable, ContextManager, Optional, TypeVar
+
+from .cache import Cache
+
+_KT = TypeVar("_KT")
+_VT = TypeVar("_VT")
+_T = TypeVar("_T", bound=Callable[..., Any])
+_T_co = TypeVar("_T_co", covariant=True)
+
+def cached(cache: Cache[_KT, _VT], key: _KT = ..., lock: Optional[ContextManager[_T_co]] = ...) -> Callable[[_T], _T]: ...
```
Comment (ikonst) [inline third_party/2and3/cachetools/decorators.pyi:10]: @hauntsaninja shouldn't the `key` parameter be typed as a `Callable[..., _KT]`?

**Pull request merged**